lib/optimizely/event_dispatcher.rb

```diff
@@ -28,26 +30,46 @@ class EventDispatcher
     # @api constants
     REQUEST_TIMEOUT = 10
 
+    def initialize(logger: nil, error_handler: nil)
+      @logger = logger || NoOpLogger.new
+      @error_handler = error_handler || NoOpErrorHandler.new
+    end
+
     # Dispatch the event being represented by the Event object.
     #
     # @param event - Event object
     def dispatch_event(event)
       if event.http_verb == :get
-        begin
-          HTTParty.get(event.url, headers: event.headers, query: event.params, timeout: REQUEST_TIMEOUT)
-        rescue Timeout::Error => e
-          return e
-        end
+        response = HTTParty.get(event.url, headers: event.headers, query: event.params, timeout: REQUEST_TIMEOUT)
+
       elsif event.http_verb == :post
-        begin
-          HTTParty.post(event.url,
-                        body: event.params.to_json,
-                        headers: event.headers,
-                        timeout: REQUEST_TIMEOUT)
-        rescue Timeout::Error => e
-          return e
-        end
+        response = HTTParty.post(event.url,
+                                 body: event.params.to_json,
+                                 headers: event.headers,
+                                 timeout: REQUEST_TIMEOUT)
       end
+
+      error_msg = "Event failed to dispatch with response code: #{response.code}"
+
+      case response.code
+      when 400...500
+        @logger.log(Logger::ERROR, error_msg)
+        @error_handler.handle_error(HTTPCallError.new("HTTP Client Error: #{response.code}"))
```

Comment:
As a consumer, i would be super frustrated that I have to parse the error message to determine the actual http response code... I would request that the code be added as a property of the error type.

And as a stylistic point, I'll just note that instead of handling the error here, all this code could be a single `rescue => e`. the 400/500 responses could be raised, and the timeout/standard error rescues be combined to a single one that logs the error message and invokes the handler. I think that would simplify all this quite a bit.